File: 程序员面试金典/4.1/findWhetherExistsPath.py
from collections import deque

# BFS
class Solution:
    def findWhetherExistsPath(self, n: int, graph: List[List[int]], start: int, target: int) -> bool:
        edges = {}
        for i, j in graph:
            if i != j:
                edges.setdefault(i, set()).add(j)
        queue = deque()
        queue.append(start)
        visited = {start}
        while queue:
            u = queue.popleft()
            if u == target:
                return True
            elif u in edges:
                for v in edges[u]:
                    if v not in visited:
                        visited.add(v)
                        queue.append(v)
        return False

File: 程序员面试金典/4.1/findWhetherExistsPath.java
// BFS
class Solution {
    public boolean findWhetherExistsPath(int n, int[][] graph, int start, int target) {
        Map<Integer, Set<Integer>> edges = new HashMap<>();
        for (int[] edge : graph)
            if (edge[0] != edge[1]) {
                if (!edges.containsKey(edge[0]))
                    edges.put(edge[0], new HashSet<>());
                edges.get(edge[0]).add(edge[1]);
            }
        Queue<Integer> queue = new LinkedList<>();
        queue.add(start);
        Set<Integer> visited = new HashSet<>();
        while (!queue.isEmpty()) {
            int v = queue.remove();
            if (v == target)
                return true;
            else if (edges.containsKey(v))
                for (int u : edges.get(v))
                    if (!visited.contains(u)) {
                        visited.add(u);
                        queue.add(u);
                    }
        }
        return false;
    }
}
